staging/src/k8s.io/kube-aggregator/pkg/controllers/autoregister/autoregister_controller.go

```diff
@@ -169,29 +197,61 @@ func (c *autoRegisterController) processNextWorkItem() bool {
 	return true
 }
 
-func (c *autoRegisterController) checkAPIService(name string) error {
+// checkAPIService syncs the current APIService against a list of desired APIService objects
+//
+// Current:           \          Desired: | A. not found | B. sync on start | C. sync always
+// ---------------------------------------|--------------|------------------|---------------
+// 1. error                               | error        | error            | error
+// 2. not found                           | -            | create once      | create
+// 3. no sync                             | -            | -                | -
+// 4. sync on start, not present at start | -            | -                | -
+// 5. sync on start, present at start     | delete once  | update once      | update once
+// 6. sync always                         | delete       | update once      | update
+func (c *autoRegisterController) checkAPIService(name string) (err error) {
 	desired := c.GetAPIServiceToSync(name)
 	curr, err := c.apiServiceLister.Get(name)
 
+	// if we've never synced this service successfully, record a successful sync.
+	hasSynced := c.hasSyncedSuccessfully(name)
+	if !hasSynced {
+		defer func() {
+			if err == nil {
+				c.setSyncedSuccessfully(name)
+			}
+		}()
+	}
+
 	switch {
-	// we had a real error, just return it
+	// we had a real error, just return it (1A,1B,1C)
 	case err != nil && !apierrors.IsNotFound(err):
 		return err
 
-	// we don't have an entry and we don't want one
+	// we don't have an entry and we don't want one (2A)
 	case apierrors.IsNotFound(err) && desired == nil:
 		return nil
 
-	// we don't have an entry and we do want one
+	// the local object only wants to sync on start and has already synced (2B,5B,6B "once" enforcement)
+	case isAutomanagedOnStart(desired) && hasSynced:
+		return nil
+
+	// we don't have an entry and we do want one (2B,2C)
 	case apierrors.IsNotFound(err) && desired != nil:
 		_, err := c.apiServiceClient.APIServices().Create(desired)
 		return err
 
-	// we aren't trying to manage this APIService.  If the user removes the label, he's taken over management himself
-	case curr.Labels[AutoRegisterManagedLabel] != "true":
```

Comment:
This worked out nicely.  Old controllers just stop mess with them.

Reply:
exactly. I'll still pick this back to 1.7, but this at least makes upgrades from 1.7 servers with this issue leave the new servers in a good state.